Expand readme diagrams and documentation (#74)
ccao-data/model-sales-val: README.md

> ================

Reviewer: **suggestion:** Let's drop the note and WIP blockquote (L3-L5).

- [Overview](#overview)
- [Flags at a glance](#flags-at-a-glance)
- [Structure of data](#structure-of-data)

> - [Important flagging details](#important-flagging-details)

Reviewer: ```suggestion
- [Flagging details](#important-flagging-details)
```

- [AWS Glue job documentation](#aws-glue-job-documentation)
- [Next steps](#next-steps)

> # Overview

Reviewer: For consistency with your other headings:

```suggestion
## Overview
```

This repository contains code to identify and flag sales that may be non-arms-length transactions. A non-arms-length sale occurs when the buyer and seller have a relationship that might influence the transaction price, leading to a sale that doesn't reflect the true market value of the property. These sales can distort our analyses and models, since they don't adhere to the principle of an open and competitive market.

> ## Sales Validation Pipeline Run Modes

Reviewer: **nitpick:** Use consistent title capitalization. This should be:

```suggestion
## Sales validation pipeline run modes
```

> ## Sales Validation Pipeline Run Modes

Reviewer: **suggestion:** I would move this entire section under the **Flagging details** section. It's important, but the breakdown of what we're flagging and what the flags are is much _more_ important.

> The Sales Validation (Sales Val) Pipeline can be executed in three distinct run modes, depending on the state of the sales data and the specific requirements for flagging:

Reviewer: **nitpick:** Sales Val Pipeline is pretty clunky. Let's do what we do in the residential and condo model README i.e.:

```suggestion
The Sales Validation Pipeline (hereafter referred to as “the pipeline”) can be executed in three distinct run modes, depending on the state of the sales data and the specific requirements for flagging:
```

1. **Initial Run:** This mode is triggered when no sales have been flagged. It's the first step in the pipeline to instantiate tables and flag sales.
2. **Glue Job:** This mode applies when there are already flagged sales in the system. It's an automated scheduled job that flags new, unflagged sales.

> 3. **Manual Update:** This mode is used when sales need to be re-flagged, either due to errors or methodology updates. This allows for the selective re-flagging of sales.

Reviewer: **praise:** Nice, this is much clearer.

### Initial Run Mode

This is the starting point for the Sales Val Pipeline when no sales are flagged.

> ```mermaid

Reviewer: **suggestion:** Let's just consolidate the diagrams into a single diagram (so the diagrams are side-by-side).

```mermaid
graph TD
    A[Schedule triggers glue job]
    B{{"Some sales are already flagged"}}
    C[Ingest data for unflagged sales]
    D[Run flagging model within glue job]
    E[Write sales data to sale.flag]
    F[Join flags to<br>default.vw_pin_sale]
    G[Save results to S3 with unique run ID]

    A -->|Trigger| B
    B -->|Process new sales| C
    C -->|Run model| D
    D -->|Output flags| E
    E -->|Join data| F
    F -->|Persist results| G
```

### Manual Update Mode

This mode allows for the manual re-flagging of sales due to errors or when an update in the methodology is required.

```mermaid
graph TD
    A{{"Sales must be re-flagged"}}
    B[Specify subset in yaml]
    C[Run manual_update.py]
    D[Increment version if sale already flagged]
    E[Assign Version = 1 if sale unflagged]
    F[Update flags in default.vw_pin_sale]
    G[Save results to S3 with new run ID]

    A -->|Manual selection| B
    B -->|Run update| C
    C -->|Version check| D
    D -->|Update process| F
    C -->|New flag| E
    E -->|Update process| F
    F -->|Persist results| G
```

## Flags at a Glance

> Sales from 2014 - present have been processed using our sales validation program. We flag properties from residential and condo data with the following specifications:

Reviewer: **suggestion:** I would add a note here about what you mean by "exclusions".

Author: I added some context here but I could use your help with it. Not sure exactly what to add. @dfsnow

<!--
/*
This query is used to generate the total sales that have some sort of outlier classification
/*

WITH TotalRecords AS (
    SELECT COUNT(*) as total_count
    FROM sale.flag
), NotOutlierCount AS (
    SELECT COUNT(*) as not_outlier_count
    FROM sale.flag
    WHERE sv_outlier_type <> 'Not outlier'
)

SELECT
    ROUND(
        (not_outlier_count * 100.0) / total_count,
        3
    ) AS not_outlier_percentage
FROM
    TotalRecords, NotOutlierCount;
-->

Around **7.2%** of the total sales have some sort of `Outlier` classification.  Within that 7.2%, the makeup of the outlier distribution is approximately as follows:

<!--
/*
This query is used to generate the proportion of different outlier types 
/*

WITH TotalRecords AS (
    SELECT COUNT(*) as total_count
    FROM sale.flag
    WHERE sv_outlier_type <> 'Not outlier'
)

SELECT 
    sv_outlier_type, 
    ROUND(COUNT(*) * 1.0 / total_count, 3) as proportion
FROM 
    sale.flag 
CROSS JOIN 
    TotalRecords
WHERE 
    sv_outlier_type <> 'Not outlier'
GROUP BY 
    sv_outlier_type, total_count
ORDER BY 
    proportion DESC;
-->

|Outlier Type     |Proportion|
|-----------------------|----------|

> |PTAX-203 flag          |0.401     |

Reviewer: **nitpick:** These should be percentages.

================

- [Overview](#overview)

> - [Flags at a Glance](#flags-at-a-glance)

Reviewer: **Nitpick:** For the sake of consistency with the other headings:

```suggestion
- [Flags at a glance](#flags-at-a-glance)
```

- [Structure of data](#structure-of-data)
- [Important flagging details](#important-flagging-details)
- [AWS Glue job documentation](#aws-glue-job-documentation)

> - [Next Steps](#next-steps)

Reviewer: ```suggestion
- [Next steps](#next-steps)
```

* Next, `glue/sales_val_flagging.py` flags all new, unflagged sales. This script is automated such that it runs on a schedule (e.g. monthly).

> * If an error occurs or we want to update the methodology on previously-flagged sales, `manual_flagging/manual_update.py` is used to select a subset of sales to re-flag. All sales have a version number that is incremented on update. When utilizing our sales views, we pull the flag data with the highest version value to keep it up-to-date.

Reviewer: **issue:** I would reframe these as the 3 possible run modes for the sales val pipeline, based on the following conditions:

1. No sales are flagged --> initial run
2. Some sales are already flagged --> glue job
3. Sales must be re-flagged --> manual update

Then update the diagrams accordingly.

> #### Local Flagging

Reviewer: **Nitpick:** Add spaces after markdown headings, also consistent capitalization:

```suggestion
#### Local flagging

```

    G -- Results saved to S3<br>with new run ID --> I

```
#### Glue Job
And here we can see how the recurrent glue job will process unflagged sales:

```mermaid

graph TD

    A[Schedule for glue job triggers a run]
    B{{Ingest data needed to flag unflagged sales}}
    C[Run flagging model within glue job]

    D[Write sales data to sale.flag]
    E[Flags joined to<br>default.vw_pin_sale]

    A -- Some sales are not flagged --> B
    B --> C
    C --> D
    D -- Results saved to S3<br>with unique run ID --> E

```

# Flags at a Glance

> Sales from 2014 - present have been processed using our sales validation program. We assign flags to the following classes of properties:

Reviewer: **issue:** Add note about not flagging certain residential sales (multi-PIN, multi-card, etc.).

- **Home flip sale (high)**: Short-term owner < 1 year & [1 high statistical outlier type]
- **Family sale (high)**: Last name match & [1 high statistical outlier type]
- **Non-person sale (high)**: Legal / corporate entity + [1 high statistical outlier type]
- **Anomaly (High)**: Anomaly algorithm (high) & [1 high statistical outlier type]
- **High price (raw & sqft )**: High price & high price per sq. ft. 
- **High price swing**: Large swing away from mean  & high price outlier 
- **High price (raw)**: High price 
- **High price (per sqft)**: High price per sq. ft. 

### Low Price

- **PTAX outlier (high)**: PTAX flag & [1 low statistical outlier type]
- **Home flip sale (low)**: Short-term owner < 1 year & [1 low statistical outlier type]
- **Family sale (low)**: Last name match & [1 low statistical outlier type]
- **Non-person sale (low)**: Legal / corporate entity + [1 low statistical outlier type]
- **Anomaly**: - Anomaly algorithm (low) & [1 low statistical outlier type]
- **Low price (raw & sqft )**: Large swing away from mean & low price outlier 
- **Low price swing**: Low price & low price per sq. ft. 
- **Low price (raw)**: Low price (or under $10k)

> - **Low price (per sqft)**: Low price per sq. ft.

Reviewer: **suggestion:** I recommend you convert all of this into a markdown table. Right now it's difficult to read.

## Distribution of Outlier Types

> Around **7.2%** of the total sales have some sort of `Outlier` classification.  Within that 7.2%, the makeup of the outlier distribution is approximately as follows:

Reviewer: **issue:** This number should be dynamically updated. See note below.

Author: Added query in html comment.

|Outlier Type     |Proportion|
|-----------------------|----------|
|PTAX-203 flag          |0.401     |
|Non-person sale (low)  |0.176     |
|Non-person sale (high) |0.081     |
|Anomaly (high)         |0.063     |
|High price (raw)       |0.058     |
|Low price (raw)        |0.054     |
|Low price (raw & sqft) |0.052     |
|Low price (sqft)       |0.021     |
|Anomaly (low)          |0.02      |
|High price (sqft)      |0.019     |
|Home flip sale (high)  |0.019     |
|High price (raw & sqft)|0.016     |
|Home flip sale (low)   |0.014     |
|Family sale (low)      |0.006     |
|Family sale (high)     |0.001     |
|Low price swing        |~0.0       |

> |High price swing       |~0.0       |

Reviewer: **issue:** If you're going to include a table like this in the README, I recommend you also include the code required to rebuild it, since this could change significantly if we add new types, alter bounds, etc.

It might be prudent to create a Quarto README that has README.md as an output. You can see an example of this in any of the R packages i.e. https://github.com/ccao-data/assessr/blob/master/README.Rmd.

Note that you can use Python in a quarto doc, no need to build the table using R.

Author: Added query in html comment.

# Important flagging details

## Sales validation pipeline run modes

The Sales Validation Pipeline (hereafter referred to as “the pipeline”) can be executed in three distinct run modes, depending on the state of the sales data and the specific requirements for flagging:

1. **Initial Run:** This mode is triggered when no sales have been flagged. It's the first step in the pipeline to instantiate tables and flag sales.
2. **Glue Job:** This mode applies when there are already flagged sales in the system. It's an automated scheduled job that flags new, unflagged sales.
3. **Manual Update:** This mode is used when sales need to be re-flagged, either due to errors or methodology updates. This allows for the selective re-flagging of sales.

```mermaid

> graph TD

Reviewer: TIL that mermaid can use subgraphs...

        C3 -->|New flag| E3
        E3 -->|Update process| F3
        F3 -->|Persist results| G3
    end

    subgraph "Glue Job Mode"
        A2[Schedule triggers glue job]
        B2{{"Some sales are already flagged"}}
        C2[Ingest data for unflagged sales]
        D2[Run flagging model within glue job]
        E2[Write sales data to sale.flag]
        F2[Join flags to<br>default.vw_pin_sale]
        G2[Save results to S3 with unique run ID]

        A2 -->|Trigger| B2
        B2 -->|Process new sales| C2
        C2 -->|Run model| D2
        D2 -->|Output flags| E2
        E2 -->|Join data| F2
        F2 -->|Persist results| G2
    end

    subgraph "Initial Run Mode"
        A1{{"No sales are flagged"}}
        B1[Run initial_flagging.py]
        C1[Flag sales as outliers or non-outliers<br>with Version = 1]
        D1[Join flags to<br>default.vw_pin_sale]
        E1[Save results to S3 with unique run ID]

        A1 -->|Initial setup| B1
        B1 -->|Flag sales| C1
        C1 -->|Store flags| D1
        D1 -->|Persist results| E1
    end

    style A1 fill:#f9f,stroke:#333,stroke-width:2px
    style B2 fill:#bbf,stroke:#333,stroke-width:2px

>     style A3 fill:#fbf,stroke:#333,stroke-width:2px

Reviewer: **issue:** These styles aren't readable (text is much too light compared to the fill color). Also, why have different colors for each subgraph?

Author: I standardized the colors for the subgraphs, but I'm not sure what you mean about the text being light. Seems like the text is all black upon rendering? 

Do you mean that the fill colors are too dark for the text?

    - Once your PR is approved, merge it into `main`. A GitHub Actions workflow called `cleanup-terraform` will delete the staging resources that were created for your branch, while a separate `deploy-terraform` run will deploy your changes to the production job and its associated resources.
  
# Next steps

## iasWorld Integration

We plan to integrate this data into our iasWorld database, so that these flags are widely available in the CCAO. We plan to upload 4-5 columns:
- Is non-arms-length (boolean)
- Flag reason (text)
- Flag run ID (hidden)
- Analyst override (boolean)
- Analyst Notes (text)

```mermaid

graph TD
    A[Data Team Data Lake] --> C1[Is non-arms-length]
    A --> C2[Flag reason]
    A --> C3[Flag run ID]
    A --> C4[Analyst override]
    
    C1 --> J[Join on doc_no]
    C2 --> J
    C3 --> J
    C4 --> J
    J --> D[iasWorld Sales Data]

    class C1 red;
    class C2 red;
    class C3 red;
    class C4 red;

```

## Analyst Review

In the future, it is our goal that human sales validation analysts will be able to review these flags and manually override them. In the longer-term future we could train a supervised model using the analysts' judgement.

> 

Reviewer: **nitpick:** Lots of unnecessary whitespace.
